jupyter_server/services/kernels/connection/channels.py

```diff
@@ -0,0 +1,809 @@
+                return
+
+            # check for closed zmq socket
+            if control_channel.closed():
+                self.log.debug("Nudge: cancelling on closed zmq socket: %s", self.kernel_id)
+                finish()
+                return
+
+            if not both_done.done():
+                log = self.log.warning if count % 10 == 0 else self.log.debug
+                log(f"Nudge: attempt {count} on kernel {self.kernel_id}")
+                self.session.send(shell_channel, "kernel_info_request")
+                self.session.send(control_channel, "kernel_info_request")
+                nonlocal nudge_handle  # type: ignore[misc]
+                nudge_handle = loop.call_later(0.5, nudge, count)
+
+        nudge_handle = loop.call_later(0, nudge, count=0)
+
+        # resolve with a timeout if we get no response
+        future = gen.with_timeout(loop.time() + self.kernel_info_timeout, both_done)
+        # ensure we have no dangling resources or unresolved Futures in case of timeout
+        future.add_done_callback(finish)
+        return _ensure_future(future)
+
+    async def _register_session(self):
+        """Ensure we aren't creating a duplicate session.
+
+        If a previous identical session is still open, close it to avoid collisions.
+        This is likely due to a client reconnecting from a lost network connection,
+        where the socket on our side has not been cleaned up yet.
+        """
+        self.session_key = f"{self.kernel_id}:{self.session.session}"
+        stale_handler = self._open_sessions.get(self.session_key)
+        if stale_handler:
+            self.log.warning("Replacing stale connection: %s", self.session_key)
+            await stale_handler.close()
+        if (
+            self.kernel_id in self.multi_kernel_manager
+        ):  # only update open sessions if kernel is actively managed
+            self._open_sessions[self.session_key] = self.websocket_handler
+
+    async def prepare(self):
+        # check session collision:
+        await self._register_session()
+        # then request kernel info, waiting up to a certain time before giving up.
+        # We don't want to wait forever, because browsers don't take it well when
+        # servers never respond to websocket connection requests.
+
+        if hasattr(self.kernel_manager, "ready"):
+            ready = self.kernel_manager.ready
+            if not isinstance(ready, asyncio.Future):
+                ready = asyncio.wrap_future(ready)
+            try:
+                await ready
+            except Exception as e:
+                self.kernel_manager.execution_state = "dead"
+                self.kernel_manager.reason = str(e)
+                raise web.HTTPError(500, str(e)) from e
+
+        t0 = time.time()
+        while not await ensure_async(self.kernel_manager.is_alive()):
+            await asyncio.sleep(0.1)
+            if time.time() - t0 > self.multi_kernel_manager.kernel_info_timeout:
```

Comment:
I think this needs parens around the left hand side of the comparison operator

Reply:
Ah right :+1: Done.